<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.33">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>guerrilla_analytics_framework – quarto_qmd</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-07ba0ad10f5680c660e360ac31d2f3b6.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-eb388a9cc6c2f7f6096b3cbc44ccb9e4.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">quarto_qmd</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./about.html" aria-current="page"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./003_Guerrilla_analytics_framework.html">Guerrilla analytics</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./001_CV.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">CV</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./002_Looking_ahead.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Looking ahead</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./003_Guerrilla_analytics_framework.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Guerrilla analytics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./004_C_elegans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">C elegans</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./005_open_peer_review.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Open peer review</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./006_relational_databases.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Relational databases</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./007_R_Package.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R package</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./008_Zotero.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Zotero</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./009_Parameterized_report.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Parameterized report</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./010_ML_Getting_started.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Getting started with machine learning</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#guerrilla-analytics" id="toc-guerrilla-analytics" class="nav-link active" data-scroll-target="#guerrilla-analytics">Guerrilla analytics</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block"></header>




<section id="guerrilla-analytics" class="level1">
<h1>Guerrilla analytics</h1>
<p>Guerrilla Analytics is data analytics performed in a very dynamic project environment that presents the team with varied and frequent disruptions and constrains the team in terms of the resources they can bring to bear on their analytics problem. <br> From: <a href="https://www.sciencedirect.com/science/article/abs/pii/B9780128002186000011">Guerrilla Analytics, 2015</a></p>
<p><br>Data science projects can get big, and keeping all files and folders tidy makes sure that it will stay possible to go through all the files and see how the analyses are performed. Keeping a well managed structure also allows for easier reproduction of the performed analyses when the folders are sent to other data scientist. <br> <br>First we take a look at what a general folder structure might look like after a few weeks of assignments:</p>
<div class="cell">
<style type="text/css">
.FixedHeightContainer
{
  height: 700px;
  padding: 3px; 
  background: rgba(0, 0, 0, .1);
  /* background: #fafafa; */
}
.Content
{
  height: 624px;
  overflow: scroll;
  background: rgba(0, 0, 0, .1);
  /*background: #fafafa;*/
}
::-webkit-scrollbar {
  -webkit-appearance: none;
  width: 7px;
}

::-webkit-scrollbar-thumb {
  border-radius: 4px;
  background-color: rgba(0, 0, 0, .5);
  box-shadow: 0 0 1px rgba(255, 255, 255, .5);
}
</style>
</div>
<div class="FixedHeightContainer">
<h3 class="anchored" data-anchor-id="guerrilla-analytics">
Old folder structure:
</h3>
<div class="Content">
<p>fs::dir_tree(path = “/Users/Alex/Documents/daur2/daurII_students_OLD”) <br> /Users/Alex/Documents/daur2/daurII_students_OLD <br> ├── Alex_Groot copy.csv <br> ├── Alex_Groot_1730201_Capstone_Exam_ABDS_DAUR2_BIOC.Rmd <br> ├── Alex_Groot_1730201_Capstone_Exam_ABDS_DAUR2_BIOC.html <br> ├── Andy.csv <br> ├── David.csv <br> ├── EXAM <br> │ ├── capstone_exam.Rmd <br> │ ├── capstone_exam.html <br> │ └── capstone_exam.md <br> ├── GDS858.soft.gz <br> ├── GPL11154.soft <br> ├── GPL96.soft <br> ├── GSE116936 <br> │ └── GSE116936_RAW.tar <br> ├── GSE116936_series_matrix.txt.gz <br> ├── GSE116936_supp_files <br> │ ├── GSM3264576_A1-Ctrl.txt.gz <br> │ ├── GSM3264577_A2-Ctrl.txt.gz <br> │ ├── GSM3264578_A3-Ctrl.txt.gz <br> │ ├── GSM3264579_B1-PGE2.txt.gz <br> │ ├── GSM3264580_B2-PGE2.txt.gz <br> │ ├── GSM3264581_B3-PGE2.txt.gz <br> │ ├── GSM3264582_I1-TNF.txt.gz <br> │ ├── GSM3264583_I2-TNF.txt.gz <br> │ ├── GSM3264584_I3-TNF.txt.gz <br> │ ├── GSM3264585_J1-TNF+PGE2.txt.gz <br> │ ├── GSM3264586_J2-TNF+PGE2.txt.gz <br> │ └── GSM3264587_J3-TNF+PGE2.txt.gz <br> ├── GSE148829 <br> │ ├── GSE148829_BEAS_Basal_Pops_TPM.txt.gz <br> │ ├── GSE148829_Human1_Basal_Pops_TPM.txt.gz <br> │ ├── GSE148829_Human2_Basal_Pops_TPM.txt.gz <br> │ ├── GSE148829_Human_Ileum_absorptiveAndCryptentero_dge.csv.gz <br> │ ├── GSE148829_Human_lung_epithelial_cell_raw_counts.txt.gz <br> │ ├── GSE148829_Mouse_Basal_Pops_TPM.txt.gz <br> │ ├── GSE148829_NHP_TB_Grans_and_Lung_dge.txt.gz <br> │ ├── GSE148829_NHP_ileum_epithelial_raw_expression_counts.csv.gz <br> │ ├── GSE148829_NHP_lung_epithelial_raw_expression_counts.csv.gz <br> │ ├── GSE148829_NasalSSS_cleaned_raw_expression_counts.csv.gz <br> │ ├── GSE148829_RAW <br> │ │ ├── GSM4487938_Mouse_Nasal_IFNa_NH11_dge.txt.gz <br> │ │ ├── GSM4487939_Mouse_Nasal_IFNa_NH12_dge.txt.gz <br> │ │ ├── GSM4487940_Mouse_Nasal_IFNa_NH21_dge.txt.gz <br> │ │ ├── GSM4487941_Mouse_Nasal_IFNa_NH22_dge.txt.gz <br> │ │ ├── GSM4487942_Mouse_Nasal_IFNa_NS11_dge.txt.gz <br> │ │ ├── GSM4487943_Mouse_Nasal_IFNa_NS12_dge.txt.gz <br> │ │ ├── GSM4487944_Mouse_Nasal_IFNa_NS21_dge.txt.gz <br> │ │ ├── GSM4487945_Mouse_Nasal_IFNa_NS22_dge.txt.gz <br> │ │ ├── GSM4488219_Mouse_basal_stim_IFNA_0.1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488220_Mouse_basal_stim_IFNA_0.1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488221_Mouse_basal_stim_IFNA_0.1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488222_Mouse_basal_stim_IFNA_0.5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488223_Mouse_basal_stim_IFNA_0.5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488224_Mouse_basal_stim_IFNA_0.5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488225_Mouse_basal_stim_IFNB_0.1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488226_Mouse_basal_stim_IFNB_0.1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488227_Mouse_basal_stim_IFNB_0.1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488228_Mouse_basal_stim_IFNB_0.5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488229_Mouse_basal_stim_IFNB_0.5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488230_Mouse_basal_stim_IFNB_0.5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488231_Mouse_basal_stim_IFNA_1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488232_Mouse_basal_stim_IFNA_1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488233_Mouse_basal_stim_IFNA_1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488234_Mouse_basal_stim_IFNA_2_REP1_TPM.txt.gz <br> │ │ ├── GSM4488235_Mouse_basal_stim_IFNA_2_REP2_TPM.txt.gz <br> │ │ ├── GSM4488236_Mouse_basal_stim_IFNA_2_REP3_TPM.txt.gz <br> │ │ ├── GSM4488237_Mouse_basal_stim_IFNB_1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488238_Mouse_basal_stim_IFNB_1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488239_Mouse_basal_stim_IFNB_1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488240_Mouse_basal_stim_IFNB_2_REP1_TPM.txt.gz <br> │ │ ├── GSM4488241_Mouse_basal_stim_IFNB_2_REP2_TPM.txt.gz <br> │ │ ├── GSM4488242_Mouse_basal_stim_IFNB_2_REP3_TPM.txt.gz <br> │ │ ├── GSM4488243_Mouse_basal_stim_IFNA_5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488244_Mouse_basal_stim_IFNA_5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488245_Mouse_basal_stim_IFNA_5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488246_Mouse_basal_stim_IFNA_10_REP1_TPM.txt.gz <br> │ │ ├── GSM4488247_Mouse_basal_stim_IFNA_10_REP2_TPM.txt.gz <br> │ │ ├── GSM4488248_Mouse_basal_stim_IFNA_10_REP3_TPM.txt.gz <br> │ │ ├── GSM4488249_Mouse_basal_stim_IFNB_5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488250_Mouse_basal_stim_IFNB_5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488251_Mouse_basal_stim_IFNB_5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488252_Mouse_basal_stim_IFNB_10_REP1_TPM.txt.gz <br> │ │ ├── GSM4488253_Mouse_basal_stim_IFNB_10_REP2_TPM.txt.gz <br> │ │ ├── GSM4488254_Mouse_basal_stim_IFNB_10_REP3_TPM.txt.gz <br> │ │ ├── GSM4488255_Mouse_basal_stim_IFNG_0.1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488256_Mouse_basal_stim_IFNG_0.1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488257_Mouse_basal_stim_IFNG_0.1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488258_Mouse_basal_stim_IFNG_0.5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488259_Mouse_basal_stim_IFNG_0.5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488260_Mouse_basal_stim_IFNG_0.5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488261_Mouse_basal_stim_IFNG_1_REP1_TPM.txt.gz <br> │ │ ├── GSM4488262_Mouse_basal_stim_IFNG_1_REP2_TPM.txt.gz <br> │ │ ├── GSM4488263_Mouse_basal_stim_IFNG_1_REP3_TPM.txt.gz <br> │ │ ├── GSM4488264_Mouse_basal_stim_IFNG_2_REP1_TPM.txt.gz <br> │ │ ├── GSM4488265_Mouse_basal_stim_IFNG_2_REP2_TPM.txt.gz <br> │ │ ├── GSM4488266_Mouse_basal_stim_IFNG_2_REP3_TPM.txt.gz <br> │ │ ├── GSM4488267_Mouse_basal_stim_IFNG_5_REP1_TPM.txt.gz <br> │ │ ├── GSM4488268_Mouse_basal_stim_IFNG_5_REP2_TPM.txt.gz <br> │ │ ├── GSM4488269_Mouse_basal_stim_IFNG_5_REP3_TPM.txt.gz <br> │ │ ├── GSM4488270_Mouse_basal_stim_IFNG_10_REP1_TPM.txt.gz <br> │ │ ├── GSM4488271_Mouse_basal_stim_IFNG_10_REP2_TPM.txt.gz <br> │ │ ├── GSM4488272_Mouse_basal_stim_IFNG_10_REP3_TPM.txt.gz <br> │ │ ├── GSM4488273_Mouse_basal_stim_Untreated_0_REP1_TPM.txt.gz <br> │ │ ├── GSM4488274_Mouse_basal_stim_Untreated_0_REP2_TPM.txt.gz <br> │ │ ├── GSM4488275_Mouse_basal_stim_Untreated_0_REP3_TPM.txt.gz <br> │ │ ├── GSM4488276_Mouse_basal_stim_Untreated_0_REP4_TPM.txt.gz <br> │ │ ├── GSM4488277_Mouse_basal_stim_Untreated_0_REP5_TPM.txt.gz <br> │ │ ├── GSM4488278_Mouse_basal_stim_Untreated_0_REP6_TPM.txt.gz <br> │ │ ├── GSM4488279_Mouse_basal_stim_Untreated_0_REP7_TPM.txt.gz <br> │ │ ├── GSM4488280_Mouse_basal_stim_Untreated_0_REP8_TPM.txt.gz <br> │ │ ├── GSM4488281_Mouse_basal_stim_Untreated_0_REP9_TPM.txt.gz <br> │ │ ├── GSM4488282_Mouse_basal_stim_Untreated_0_REP10_TPM.txt.gz <br> │ │ ├── GSM4488283_Mouse_basal_stim_Untreated_0_REP11_TPM.txt.gz <br> │ │ └── GSM4488284_Mouse_basal_stim_Untreated_0_REP12_TPM.txt.gz <br> │ └── GSE148829_RAW.tar <br> ├── GSE75073 <br> │ ├── GSE75073_RAW.tar <br> │ ├── GSE75073_Signals_450k.txt.gz <br> │ └── GSE75073_Signals_Epic.txt.gz <br> ├── GSE76073 <br> │ ├── GSE76073_Normalized_counts.txt.gz <br> │ └── GSE76073_Raw_counts.txt.gz <br> ├── GSE923_series_matrix.txt.gz <br> ├── Images <br> │ ├── 2020-04-20CORT.png <br> │ ├── 20200403_citrul.png <br> │ ├── 20200404_citrul.png <br> │ ├── 20200404_ifabp.png <br> │ ├── Bioconductor.png <br> │ ├── Biostrings.png <br> │ ├── BrowseSeq <br> │ │ ├── Dia1.PNG <br> │ │ └── sequence_alignment_plots <br> │ │ ├── Dia1.PNG <br> │ │ ├── Dia2.PNG <br> │ │ ├── Dia3.PNG <br> │ │ ├── Dia4.PNG <br> │ │ ├── Dia5.PNG <br> │ │ ├── Dia6.PNG <br> │ │ ├── Dia7.PNG <br> │ │ └── Dia8.PNG <br> │ ├── DESeq2.png <br> │ ├── Dia1.JPG <br> │ ├── Dia3.JPG <br> │ ├── GenomicRanges.png <br> │ ├── Git_Setup_1.JPG <br> │ ├── Glimma.png <br> │ ├── IRanges.png <br> │ ├── MSnbase.png <br> │ ├── Project_1.jpg <br> │ ├── Project_2.jpg <br> │ ├── Rlogo.png <br> │ ├── Rtools.png <br> │ ├── bioconductorlogo.jpg <br> │ ├── cat_sleeping_on_keyboard.jpg <br> │ ├── citrulline.png <br> │ ├── connectome-1.png <br> │ ├── cortline.png <br> │ ├── covid19_ziekenhuis.png <br> │ ├── dotdotdot_rotate_axis.png <br> │ ├── edgeR.png <br> │ ├── exercise_2_1_1.png <br> │ ├── exercise_2_1_1_c.png <br> │ ├── forcats.png <br> │ ├── git_web.png <br> │ ├── gitinstallermac.png <br> │ ├── heatmap.png <br> │ ├── heatmap1.png <br> │ ├── heatmap2.png <br> │ ├── hex-ggplot2.png <br> │ ├── installpackages.png <br> │ ├── loadpackages.png <br> │ ├── logo_R.jpg <br> │ ├── ma1.png <br> │ ├── ma2.png <br> │ ├── mzR_hl.png <br> │ ├── new_proj_done.png <br> │ ├── new_proj_git.png <br> │ ├── one-doesnot-simply.jpg <br> │ ├── open_486.jpg <br> │ ├── open_compaq.jpg <br> │ ├── pca1.png <br> │ ├── pepper.jpg <br> │ ├── pipe.png <br> │ ├── purrr_small.png <br> │ ├── purrr_sticker.jpg <br> │ ├── pvalues.png <br> │ ├── r4ds.png <br> │ ├── readr_sticker.png <br> │ ├── readxl_sticker.png <br> │ ├── rmac_download.png <br> │ ├── rstudio_download.png <br> │ ├── rstudio_leeg.png <br> │ ├── rwin_download.png <br> │ ├── security.png <br> │ ├── setup-wizard-rstudio.png <br> │ ├── setupRtools.png <br> │ ├── stringr.png <br> │ ├── summarizedexperiment.png <br> │ ├── tidyr_sticker.png <br> │ ├── tidyverse_sticker.png <br> │ ├── usrlocal.png <br> │ ├── wizardRmac.png <br> │ ├── wizardRwindows.png <br> │ ├── wontinstall.png <br> │ ├── workflow_course.png <br> │ ├── xquarkdisk.png <br> │ ├── xquarkwizard.png <br> │ ├── xquartz.png <br> │ └── yoda_the_last_jedi.jpg <br> ├── John.csv <br> ├── Les 6.Rmd <br> ├── Les4.Rmd <br> ├── Les4.nb.html <br> ├── Les5.Rmd <br> ├── Lesson_5_Exercise_a1.Rmd <br> ├── Mike.csv <br> ├── R <br> │ └── do_tidy_pertussis.R <br> ├── README.md <br> ├── Steve.csv <br> ├── answers <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment.Rmd <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment.html <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment_files <br> │ │ └── figure-html <br> │ │ ├── unnamed-chunk-23-1.png <br> │ │ ├── unnamed-chunk-24-1.png <br> │ │ ├── unnamed-chunk-29-1.png <br> │ │ └── unnamed-chunk-30-1.png <br> │ ├── daur2_lesson_6_exercise_rnaseq.Rmd <br> │ ├── daur2_lesson_6_exercise_rnaseq.html <br> │ ├── daur2_lesson_6_exercise_rnaseq_student_version.Rmd <br> │ └── daur2_lesson_6_exercise_rnaseq_student_version.html <br> ├── caspase_align.aux <br> ├── caspase_align.log <br> ├── caspase_align.tex <br> ├── code <br> │ └── load_data.R <br> ├── data <br> │ ├── CISID_pertussis_10082018.csv <br> │ ├── biomaRt_homology_example.rds <br> │ ├── biomaRt_homology_example_aa.rds <br> │ ├── biomaRt_homology_example_dna.rds <br> │ ├── chircus_blg_aminoacids.fasta <br> │ ├── course_r_packages.tsv <br> │ ├── course_r_packages.txt <br> │ ├── covid_rivm <br> │ ├── go_bp-hit.csv <br> │ ├── go_cc-hit.csv <br> │ ├── go_mf-hit.csv <br> │ ├── hobbit_chapter.txt <br> │ ├── hobbit_dwarves.txt <br> │ ├── hox_aa_sequences.txt <br> │ ├── hox_dna_sequences.txt <br> │ ├── human_hoxb1_protein.fasta <br> │ ├── human_hoxb1_protein.gb <br> │ ├── human_rhodopsin_protein.fasta <br> │ ├── human_ttn_genomic_dna.fasta <br> │ ├── jp_coldata.txt <br> │ ├── jp_counts.txt <br> │ ├── jp_features.txt <br> │ ├── jp_miame.rds <br> │ ├── jp_miame.txt <br> │ ├── lesson2 <br> │ │ ├── ENST00000642929.1.fa <br> │ │ ├── ENST00000643997.1.fa <br> │ │ ├── ENST00000644264.1.fa <br> │ │ ├── ENST00000644696.1.fa <br> │ │ ├── ENST00000644919.1.fa <br> │ │ ├── ENST00000646017.1.fa <br> │ │ ├── ENST00000646618.2.fa <br> │ │ ├── ENST00000646644.1.fa <br> │ │ ├── ENST00000647220.1.fa <br> │ │ ├── ENST00000647443.1.fa <br> │ │ ├── cholesterol.xlsx <br> │ │ ├── diet <br> │ │ │ ├── Andy.csv <br> │ │ │ ├── David.csv <br> │ │ │ ├── John.csv <br> │ │ │ ├── Mike.csv <br> │ │ │ └── Steve.csv <br> │ │ ├── diet_data.zip <br> │ │ └── sequence1.txt <br> │ └── lesson3 <br> │ ├── GO-terms.txt <br> │ ├── dna_random <br> │ ├── sequence.fasta <br> │ ├── sequence1.txt <br> │ └── sequence2.txt <br> ├── daurII_students <br> │ ├── R <br> │ │ └── do_tidy_pertussis.R <br> │ ├── README.md <br> │ ├── code <br> │ │ └── load_data.R <br> │ ├── data <br> │ │ ├── CISID_pertussis_10082018.csv <br> │ │ ├── course_r_packages.tsv <br> │ │ ├── covid_rivm <br> │ │ ├── go_bp-hit.csv <br> │ │ ├── go_cc-hit.csv <br> │ │ ├── go_mf-hit.csv <br> │ │ ├── hobbit_chapter.txt <br> │ │ ├── hobbit_dwarves.txt <br> │ │ ├── hox_aa_sequences.txt <br> │ │ ├── hox_dna_sequences.txt <br> │ │ ├── human_hoxb1_protein.fasta <br> │ │ ├── human_hoxb1_protein.gb <br> │ │ ├── human_rhodopsin_protein.fasta <br> │ │ ├── human_ttn_genomic_dna.fasta <br> │ │ └── lesson2 <br> │ │ └── diet_data.zip <br> │ ├── daurII_students.Rproj <br> │ └── demos <br> │ └── 01_demo_functions.Rmd <br> ├── daurII_students.Rproj <br> ├── demos <br> │ ├── 01_demo_functions.Rmd <br> │ ├── 021_demo_2_agenda.Rmd <br> │ ├── 021_demo_2_agenda.html <br> │ ├── 022_lesson2_map_nest.Rmd <br> │ ├── 022_lesson2_map_nest.html <br> │ ├── 031_strings_and_regex.Rmd <br> │ ├── 031_strings_and_regex.html <br> │ ├── 031_strings_and_regex.pdf <br> │ ├── 041_biomart.Rmd <br> │ ├── 042_geoquery.Rmd <br> │ ├── 043_rentrez.Rmd <br> │ ├── 051_demo_multiple_sequence_alignment.Rmd <br> │ ├── 051_demo_multiple_sequence_alignment.html <br> │ ├── 051_demo_multiple_sequence_alignment.md <br> │ ├── 051_demo_multiple_sequence_alignment_files <br> │ │ └── figure-html <br> │ │ ├── unnamed-chunk-23-1.png <br> │ │ ├── unnamed-chunk-24-1.png <br> │ │ ├── unnamed-chunk-29-1.png <br> │ │ └── unnamed-chunk-30-1.png <br> │ ├── 052_demo_summarized_experiment.Rmd <br> │ ├── 052_demo_summarized_experiment.html <br> │ ├── 052_demo_summarized_experiment.md <br> │ ├── 61_heatmaps.Rmd <br> │ ├── 61_heatmaps.html <br> │ ├── 62_rnseq_workflow.Rmd <br> │ └── 62_rnseq_workflow.html <br> ├── les1.Rmd <br> ├── les1.nb.html <br> ├── les2.Rmd <br> ├── les2.nb.html <br> ├── les3.Rmd <br> ├── my_transcript.fasta <br> ├── resources.Rmd <br> └── unzip <br> ├── Andy.csv <br> ├── David.csv <br> ├── John.csv <br> ├── Mike.csv <br> └── Steve.csv <br></p>
</div>
</div>
<br> <br> As you can see, the original file structure used during the earlier data science course is quite the mess. Instead we now make usage of the Guerilla analytics framework to store data and manage files. Using simple structures to store data, code, generated graphs and makes it easier to use and replicate analyses, even years after the analyses are performed. <br> Making sure all files in a environment pertain only to that single project is important. And to make data storage simple, a central data folder in the root of the enviornment is used with clear naming conventions, dates and versions. Also metadata files are used to allow anyone with access to the project to see what the contents of the project are. <br> <br>Here is the above file structure, with the guerilla analytics framework <b>retroactively </b>applied: <br>
<div class="FixedHeightContainer">
<h3 class="anchored">
New folder structure:
</h3>
<div class="Content">
<p>├── EXAM <br> │ ├── Alex_Groot copy.csv <br> │ ├── capstone_exam.Rmd <br> │ ├── capstone_exam.html <br> │ └── capstone_exam.md <br> ├── Images <br> │ ├── 2020-04-20CORT.png <br> │ ├── 20200403_citrul.png <br> │ ├── BrowseSeq <br> │ │ ├── Dia1.PNG <br> │ │ └── sequence_alignment_plots <br> │ │ ├── Dia1.PNG <br> │ │ ├── Dia2.PNG <br> │ └── yoda_the_last_jedi.jpg <br> ├── R <br> │ └── do_tidy_pertussis.R <br> ├── README.md <br> ├── caspase_align.aux <br> ├── caspase_align.log <br> ├── caspase_align.tex <br> ├── code <br> │ └── load_data.R <br> ├── data_raw <br> │ ├── lesson2 <br> │ │ ├── ENST00000642929.1.fa <br> │ │ ├── ENST00000643997.1.fa <br> │ │ ├── ENST00000644264.1.fa <br> │ │ ├── cholesterol.xlsx <br> │ │ ├── diet <br> │ │ │ ├── Andy.csv <br> │ │ │ └── Steve.csv <br> │ │ ├── diet_data.zip <br> │ │ └── sequence1.txt <br> │ ├── lesson3 <br> │ │ ├── GO-terms.txt <br> │ │ ├── dna_random <br> │ │ ├── sequence.fasta <br> │ │ ├── sequence1.txt <br> │ │ └── sequence2.txt <br> │ └── unsorted <br> │ ├── Andy.csv <br> │ ├── CISID_pertussis_10082018.csv <br> │ ├── David.csv <br> │ ├── GSE116936 <br> │ │ └── GSE116936_RAW.tar <br> │ ├── GSE116936_series_matrix.txt.gz <br> │ ├── GSE116936_supp_files <br> │ │ ├── GSM3264576_A1-Ctrl.txt.gz <br> │ │ ├── GSM3264577_A2-Ctrl.txt.gz <br> │ │ ├── GSM3264578_A3-Ctrl.txt.gz <br> │ │ ├── GSM3264585_J1-TNF+PGE2.txt.gz <br> │ │ ├── GSM3264586_J2-TNF+PGE2.txt.gz <br> │ │ └── GSM3264587_J3-TNF+PGE2.txt.gz <br> │ ├── GSE148829 <br> │ │ ├── GSE148829_BEAS_Basal_Pops_TPM.txt.gz <br> │ │ ├── GSE148829_Human1_Basal_Pops_TPM.txt.gz <br> │ │ ├── GSE148829_Human2_Basal_Pops_TPM.txt.gz <br> │ │ │ ├── GSM4487938_Mouse_Nasal_IFNa_NH11_dge.txt.gz <br> │ │ │ ├── GSM4487939_Mouse_Nasal_IFNa_NH12_dge.txt.gz <br> │ │ │ ├── GSM4487940_Mouse_Nasal_IFNa_NH21_dge.txt.gz <br> │ │ │ ├── GSM4487941_Mouse_Nasal_IFNa_NH22_dge.txt.gz <br> │ │ │ ├── GSM4487942_Mouse_Nasal_IFNa_NS11_dge.txt.gz <br> │ │ │ ├── GSM4487943_Mouse_Nasal_IFNa_NS12_dge.txt.gz <br> │ │ │ ├── GSM4487944_Mouse_Nasal_IFNa_NS21_dge.txt.gz <br> │ │ │ ├── GSM4487945_Mouse_Nasal_IFNa_NS22_dge.txt.gz <br> │ │ │ ├── GSM4488282_Mouse_basal_stim_Untreated_0_REP10_TPM.txt.gz <br> │ │ │ ├── GSM4488283_Mouse_basal_stim_Untreated_0_REP11_TPM.txt.gz <br> │ │ │ └── GSM4488284_Mouse_basal_stim_Untreated_0_REP12_TPM.txt.gz <br> │ │ └── GSE148829_RAW.tar <br> │ ├── GSE75073 <br> │ │ ├── GSE75073_RAW.tar <br> │ │ ├── GSE75073_Signals_450k.txt.gz <br> │ │ └── GSE75073_Signals_Epic.txt.gz <br> │ ├── GSE76073 <br> │ │ ├── GSE76073_Normalized_counts.txt.gz <br> │ │ └── GSE76073_Raw_counts.txt.gz <br> │ ├── GSE923_series_matrix.txt.gz <br> │ ├── human_hoxb1_protein.fasta <br> │ ├── human_hoxb1_protein.gb <br> │ ├── human_rhodopsin_protein.fasta <br> │ ├── human_ttn_genomic_dna.fasta <br> │ ├── jp_coldata.txt <br> │ ├── jp_counts.txt <br> │ ├── jp_features.txt <br> │ ├── jp_miame.txt <br> │ ├── my_transcript.fasta <br> │ └── unzip <br> │ ├── Andy.csv <br> │ ├── David.csv <br> │ ├── John.csv <br> │ ├── Mike.csv <br> │ └── Steve.csv <br> ├── data_rds <br> │ ├── biomaRt_homology_example.rds <br> │ ├── biomaRt_homology_example_aa.rds <br> │ ├── biomaRt_homology_example_dna.rds <br> │ └── jp_miame.rds <br> ├── daurII_students <br> │ ├── R <br> │ │ └── do_tidy_pertussis.R <br> │ ├── README.md <br> │ ├── code <br> │ │ └── load_data.R <br> │ ├── data <br> │ │ ├── CISID_pertussis_10082018.csv <br> │ │ ├── course_r_packages.tsv <br> │ │ ├── covid_rivm <br> │ │ ├── human_rhodopsin_protein.fasta <br> │ │ ├── human_ttn_genomic_dna.fasta <br> │ │ └── lesson2 <br> │ │ └── diet_data.zip <br> │ ├── daurII_students.Rproj <br> │ └── demos <br> │ └── 01_demo_functions.Rmd <br> ├── daurII_students.Rproj <br> ├── demos <br> │ ├── 01_demo_functions.Rmd <br> │ ├── 021_demo_2_agenda.Rmd <br> │ ├── 021_demo_2_agenda.html <br> │ ├── 61_heatmaps.Rmd <br> │ ├── 61_heatmaps.html <br> │ ├── 62_rnseq_workflow.Rmd <br> │ └── 62_rnseq_workflow.html <br> └── markdown <br> ├── Alex_Groot_1730201_Capstone_Exam_ABDS_DAUR2_BIOC.Rmd <br> ├── Alex_Groot_1730201_Capstone_Exam_ABDS_DAUR2_BIOC.html <br> ├── Les 6.Rmd <br> ├── Les4.Rmd <br> ├── Les4.nb.html <br> ├── Les5.Rmd <br> ├── Lesson_5_Exercise_a1.Rmd <br> ├── answers <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment.Rmd <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment.html <br> │ ├── daur2_lesson51_demo_multiple_sequence_alignment_files <br> │ │ └── figure-html <br> │ │ ├── unnamed-chunk-23-1.png <br> │ │ ├── unnamed-chunk-24-1.png <br> │ │ ├── unnamed-chunk-29-1.png <br> │ │ └── unnamed-chunk-30-1.png <br> │ ├── daur2_lesson_6_exercise_rnaseq.Rmd <br> │ ├── daur2_lesson_6_exercise_rnaseq.html <br> │ ├── daur2_lesson_6_exercise_rnaseq_student_version.Rmd <br> │ └── daur2_lesson_6_exercise_rnaseq_student_version.html <br> ├── les1.Rmd <br> ├── les1.nb.html <br> ├── les2.Rmd <br> ├── les2.nb.html <br> ├── les3.Rmd <br> └── resources.Rmd <br> <br></p>
</div>
</div>
<p><br> <br> In order to have the folder structured in the Guerrilla analytics framework method, I retroactively moved the files. This is just used as an example to show what a tidier and better organized structure might look like. The big problem with this is that it <b>will </b> break many, if not all, of the scripts inside due to changing of file paths.</p>
<p>The best way to keep files well-managed is to set up a file structure before starting and sticking to it. Keeping a folder with all the raw data with its names the way they were obtained makes sure confusion is kept to a minimum when the data is looked at. Creating seperate data folders enables to track transformed data while preserving the raw data.</p>
<p>And keeping metadata files in all the folders that explain not just folder structure, but importantly also file naming conventions and potentially used abbreviations for them will mean that even years later someone will be able to look at the project and learn what is inside.</p>
<p>This portfolio was made with the Guerrilla analytics framework in mind from the start:</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>